### DISCUSSION TITLE
Backing up resources that are not managed by Flux

### DISCUSSION THREAD

**Discussion by rombert** [fluxcd/flux2#2403] [Q&A]
Hi,

I am wondering, what are others doing for backup? The functionality that I am looking for is to back up resources not managed by Flux, mainly persistent volumes. Cert-manager certificates are a nice-to-have.

I think the most used tool is velero, but it looks like the most prevalent use case is a full backup-restore of the full cluster. Since I am using flux, I am more concerned with restoring persistent volume data, either on a case-by-case basis, or all of them altogether.

I have not yet managed to convince velero (with restic) to back up just the persistent volumes, I seem to need to include the pods and then the persistent volumes are discovered...

At any rate, I am curious about what others are doing for backing up non-flux-managed objects in their Kubernetes deployments.

Thanks!

**Comment by boedy** [comment]
@rombert A couple of months have passed since your post. Have you managed to find a workable solution?

I'm also trying to figure out how to best combine Flux with Velero. Specifically how to restore a cluster running stateful applications in case of disaster recovery.

**Comment by rombert** [comment]
@boedy - I did not make progress yet unfortunately. I've recently moved all my SQL workloads ( well, a couple of MariaDB instance ) to a MariaDB instance hosted on a VM, which already has a backup process in place. I don't have a good solution for persistent volumes though.

**Comment by Nils98Ar** [comment]
@rombert Any progress?

**Reply by rombert** [reply]
Nope, I did not look to much into it. Right now I am manually taking snapshots of the persistent volumes manually, by executing commands directly on the storage server, while the workloads are scaled down.

**Comment by juliafmorgado** [comment]
@rombert - If you're still exploring options you might find Veeam Kasten a useful alternative (not open source). Kasten is purpose-built for Kubernetes data management, offering seamless backup, restore, and mobility capabilities, including for persistent volumes and stateful applications. Additionally, Kasten integrates with open-source projects like Kanister, which provides a framework for data operations, enabling more advanced workflows such as custom data transformations during backup and restore. This combination could give you more control over selective data protection and recovery in Kubernetes environments, tailored to your needs. It might be worth exploring!

**Reply by rombert** [reply]
Thanks for the node @juliafmorgado . Looking at Kasten I noted that

> Currently, on-premises Veeam Kasten deployments are only supported if storage is provisioned via the Container Storage Interface (CSI).

 https://docs.kasten.io/latest/restrictions.html#on-premises-storage-providers  

That is unfortunately not my case right now. But I'll keep a lookout for Kasten when I'm getting back to this topic.

**Comment by svengreb** [comment]
One popular OSS project (also used in commercial products like OpenShift) is [VolSync](https://github.com/backube/volsync). It is based on the concept of "movers" that are baked by large and stable projects like [restic](https://restic.net), [rclone](https://rclone.org) (used by _restic_), [syncthing](https://syncthing.net), and good ol' [rsync](https://rsync.samba.org).
It is based on CRDs where [a backup is created with a `ReplicationSource` resource](https://volsync.readthedocs.io/en/stable/usage/restic/index.html#configuring-backup) which can be [restored with into a new volume with `ReplicationDestination` resource](https://volsync.readthedocs.io/en/stable/usage/restic/index.html#performing-a-restore). The backup target is simply the name of the volume where, based on the used _mover_, some special configurations are possible, e.g. selecting only specific paths (sadly [not yet possible](https://github.com/backube/volsync/issues/959) with _restic_ but other movers like _rclone_).

I used _VolSync_ for all of my private clusters and the plan is to introduce it to clusters at my job too. Flux is used for all clusters already and the combo of both works like a charm 😄 

Note that only [the _restic_ mover](https://volsync.readthedocs.io/en/stable/usage/restic) is a pure backup solution while all other movers can be used for backups, but are mainly used to continuously synchronize data between volumes, within the same cluster or across clusters.

**Reply by svengreb** [reply]
Another project is [k8up](https://k8up.io) which I also consider to test and maybe it even fits better for backup-only use cases while the basic way how the CRDs works are similar, but _k8up_ comes with some additional features like "Pre-Backup Pods" that allow to run custom logic before a backups starts like creating database dump to ensure data consistency.

**Reply by svengreb** [reply]
Some more results from my research: [stash](https://github.com/stashed/stash). It's build on top of a freemium model where most features for homelab and private users are included in the community edition. It comes with one (in my opinion) great features that are (sadly) still missing (but planned) in [VolSync](https://github.com/backube/volsync/issues/959) and [k8up](https://github.com/k8up-io/k8up/issues/317): pattern to [include/exclude specific files and directories](https://stash.run/docs/v2024.9.30/guides/use-cases/exclude-include-files)!
I'll also give it a try since it looks like an advanced project that should be around in the future due to the (hopefully) financial base plan, even though _k8up_ is also [a CNCF "sandbox" project](https://landscape.cncf.io/?item=runtime--cloud-native-storage--k8up) which is great for a possibly stable financial support in the future through CNCF.

**Reply by svengreb** [reply]
### Update regarding _stash_

As of 2024-01-01 [the "community" edition is not available anymore](https://appscode.com/blog/post/community-license), but [they "forgot" to update their GitHub repository README](https://github.com/stashed/stash/issues/1569). So there goes another OSS project into the lawyer hell that stands on top of first-class OSS projects like [restic](https://restic.net), but turn into money grabbers as soon as their project gets larger. Welcome to the new reality where any large OSS project should better update their license to prevent any of these companies to never be able to use their code again for all times. Sorry for the rage, it's just a pure shame for humanity that individual maintainers of essential projects (restic, curl, etc.) are left behind while others suck on the money cow.

**Reply by rombert** [reply]
Thanks for the detailed info @svengreb . I'm not looking into this topic right now but your notes are very valuable.
